### PR TITLE
Prototype: merge block CSS with `theme.json` styles

### PR DESCRIPTION
**Pull request by oandregal**
Implements https://github.com/WordPress/gutenberg/issues/33977

**I'd like feedback on the direction. Should we want to pursue this, I'd improve the implementation later**.

This PR experiments with moving parts of the CSS of the block to its `block.json` so it can be used to be merged with `theme.json` styles. The result of the merging will be (later in the chain overwrites the previous):

> core (theme.json) > block (block.json) > theme (theme.json) > user (styles coming from the global styles sidebar)

There're a few advantages to this approach:

- Block styles become easily configurable by the themes (via `theme.json`).
- We ship either block or theme styles, but not both.

I've noticed that many of the block styles are attached to classes (has-background, style variations, etc) so I presume, initially, we won't see a big reduction of CSS. It'll be a good first step to absorb more later (e.g.: [block variations](https://github.com/WordPress/gutenberg/issues/33232)).

## How to test

- Use the empty theme.
- Verify that the `global-styles-inline-css` embedded stylesheet in the frontend contains the following CSS coming from the `block.json`:

```css
.wp-block-button__link {
  background-color: #32373c;
  color: #fff;
  font-size: 1.125em;
  padding-top: calc(0.667em + 2px);
  padding-right: calc(1.333em + 2px);
  padding-bottom: calc(0.667em + 2px);
  padding-left: calc(1.333em + 2px);
}
```

- Add the following to the `theme.json` of the theme:

```json
"styles": {
  "blocks": {
    "core/button": {
      "color": {
        "text": "#000",
        "background": "#fff"
      }
    }
  }
}
```

- Verify that the `global-styles-inline-css` embedded stylesheet in the frontend contains the following CSS (colors come from the theme now):

```css
.wp-block-button__link {
  background-color: #fff;
  color: #000;
  font-size: 1.125em;
  padding-top: calc(0.667em + 2px);
  padding-right: calc(1.333em + 2px);
  padding-bottom: calc(0.667em + 2px);
  padding-left: calc(1.333em + 2px);
}
```

### CODE REVIEW

**Review by youknowriad — COMMENTED** [review]
I like this direction personally. Should we also update the site editor style rendered?
Curious to get more opinions on this one. @gziolo @mcsf @talldan @mtias

**Comment by jasmussen** [comment]
I like this principle. It does open up a bit of challenges, though, perhaps best illustrated here:

Screenshot 2021-08-23 at 10 34 44

The most important properties of the button are extracted from the style and put into the JSON file for easy parsing, tweaking and disabling at a theme's whim. This is cool, and probably sufficient. But it's obviously also not all of it, as some styles do not have JSON counterparts. It's probably fine as in most cases these are properties that are structurally sensible, such as the `cursor` property, or the inline-block `display`. The box sizing property appears to exist primarily for the button width properties to behave as intended. But it does mean we'll still likely need to keep some CSS around.

Then there are the curious bits, for example those padding values, `padding: calc(0.667em + 2px) calc(1.333em + 2px);`. They exist to make sure that the solid non-bordered button has the same inner padding as the 2px bordered button. I'm happy to see that easily ported to the json file, which should hopefully enable us to absorb these two styles as shortcuts to border/padding configurations, rather then classname applications:

Screenshot 2021-08-23 at 10 53 32

The padding challenge remains, though. If we were to absorb those styles, the outline would still need to have a `border` rule, meaning we'd still need those curious padding shenanigans in order to keep the padding intact. @kjellr @MaggieCabrera do you have any ideas for how we might refactor the buttons so that the solid and bordered versions appear equally tall? [Here's a codepen to play around in](https://codepen.io/joen/pen/VwWwZLb?editors=1100).

**Comment by MaggieCabrera** [comment]
This padding issue has been a headache for us. We've had a use case that complicates it even more, and it's a very common one for themes to implement, that is related to not having any hover states built in on the block. The case is simple: a button block that on hover looks like the outline variation. For a theme to do that it has to build in this padding trickery and then remove it when the hover happens, because then there's a border that didn't exist before. Ideally, I would like the solid button style to have a border width set to the same color as the button background but I'm guessing that won't work well for backward compatibility, will it?

@jasmussen your codepen is exactly what I would do in this case too, but the theme.json can actually define a border-width, how would that be handled then?

/cc @pbking @jffng @scruffian

**Comment by oandregal** [comment]
I'm trying this with the empty theme and the padding still behaves as expected:

Captura de ecrã de 2021-08-23 11-39-14

The reduced padding for the outline still applies, although the default padding is now part of `global-styles-inline-css` instead of the `block-library` CSS. Did I read your comment correctly or was there a different issue?

**Comment by MaggieCabrera** [comment]
Adding this to emptytheme's json:

```
	"styles": {
		"blocks": {
			"core/button": {
				"border": {
					"radius": "0",
					"color": "red",
					"style": "solid",
					"width": "6px"
				}
			}
		}		
	}
```

I get, on trunk:

Editor | Frontend
--- | ---
Screenshot 2021-08-23 at 11 54 27 | Screenshot 2021-08-23 at 11 54 23


And with this PR:
Editor | Frontend
--- | ---
Screenshot 2021-08-23 at 11 58 49 | Screenshot 2021-08-23 at 11 58 54

I don't know why border-radius is not picking up on trunk :S

**Comment by jasmussen** [comment]
> your codepen is exactly what I would do in this case too, but the theme.json can actually define a border-width, how would that be handled then?

That's exactly the headache I was hoping to find a better solution to. My custom CSS solution is usually to use an inner box-shadow instead of the outset border — `outline` isn't an option as it can't have a border radius applied. But our border control adds actual borders, not box-shadows, so in case we were to absorb the solid and outline styles as inspector control shortcuts it'd just be nice to have a neat and simple solution for the height to be the same. The calc rule is a nice hack, but as soon as a user overrides the padding on a button, mixing and matching the two styles causes drift. Maybe it's fine? 

The alternate solutions I was pondering, adding an explicit height or using pseudo elements, have their own sets of downsides.

**Comment by MaggieCabrera** [comment]
> The calc rule is a nice hack, but as soon as a user overrides the padding on a button, mixing and matching the two styles causes drift. Maybe it's fine?

I would say it's fine if it only happened when the user sets it, but not so much when it's the theme.json doing so. You are just passing the headache to the themers :D

**Comment by jasmussen** [comment]
The solid/outlined button problem is a bit tangential to the conversation here, the principle of moving these styles over still feels valid to me. This was primarily an exploration of one aspect that will start to surface as we start to do this. It's probably a good thing, as we would've encountered this same issue with the button eventually anyway.

**Comment by MaggieCabrera** [comment]
Yeah, I believe these issues are inherent to the block but shouldn't be a blocker for this particular change

**Comment by oandregal** [comment]
I'm looking at the issue with border-radius Maggie shared.

For the challenges of maintaining height: isn't the issue that the user can't modify the underlying values of the block style variations? If they could, when they change padding in the default state, it is also the user's responsibility to update it in the outline variation. Not a final solution but https://github.com/WordPress/gutenberg/issues/33232 should get us moving.

**Comment by oandregal** [comment]
This is what happens: the outline variation uses the selectors `.is-style-outline > .wp-block-button__link, .wp-block-button__link.is-style-outline`, with specificity 020.

The global styles rule for the button:

- In the frontend, uses `.wp-block-button__link`, 010, so the outline rule wins.
- In the editor, the global styles are wrapped by the editor class and the rule ends being `.editor-styles-wrapper .wp-block-button__link`, also 020. Because it's found later in the document than the outline, the global styles one wins.

I can repro the same thing in `trunk`. The fix for this is actually building https://github.com/WordPress/gutenberg/issues/33232 and that the themes can provide styles for the variations. This one is getting more urgent to implement.

**Comment by oandregal** [comment]
As for the border-radius, I didn't see any report, so filed an issue at https://github.com/WordPress/gutenberg/issues/34240

**Comment by jasmussen** [comment]
> For the challenges of maintaining height: isn't the issue that the user can't modify the underlying values of the block style variations? If they could, when they change padding in the default state, it is also the user's responsibility to update it in the outline variation. Not a final solution but #33232 should get us moving.

The challenge is that showing a solid background button next to a bordered outlined button with the same height, is remarkably harder to accomplish than you'd think. You'd think the solid button just has a background and no border, while the outlined version has a border and no background, right? Yes, but then the border is outset from the fill, causing the height differential, shown here with a 4px border:

Screenshot 2021-08-24 at 09 55 40

The way we've done it so far is to apply `padding: calc(1em + 4px);` to the solid button to make it as big as the 4px bordered button:

Screenshot 2021-08-24 at 09 58 18

But `calc(1em + 4px)` is not a very intuitive value for someone using the block editor to change. I'm not even sure we can surface it as a configurable value at all. And it certainly wouldn't be intuitive for anyone seeing to mix the two button styles to figure out why the height is off when the padding on both buttons is the same.

Another option would be to apply the outline version using `box-shadow: inset 0 0 0 4px currentColor;`. That would work fine and use the same padding on both button styles. But then we can't use the Border control inspector panel, as that draws `border`s, not `box-shadow`s — at least not yet.

A third option is to also draw a black border on the solid button, so both are the same height. But then you'd have to change both the background and the border color in order to change the color of the solid button. Which might be okay, but wouldn't work very well with gradient color buttons.

I found a third option here: https://codepen.io/joen/pen/yLXLEgW?editors=1100 — always drawing a transparent border on the solid button, that seems to get pretty close to accomplishing what we need. The challenge here would then be if you change the border width on the outline button, you'd also have to change the border width on the solid version, even if it was transparent. Not sure if that would be a good experience 🤔  

@MaggieCabrera any thoughts on that last option above? ☝️

**Comment by MaggieCabrera** [comment]
> I found a third option here: https://codepen.io/joen/pen/yLXLEgW?editors=1100 — always drawing a transparent border on the solid button, that seems to get pretty close to accomplishing what we need. The challenge here would then be if you change the border width on the outline button, you'd also have to change the border width on the solid version, even if it was transparent. Not sure if that would be a good experience 🤔
> 
> @MaggieCabrera any thoughts on that last option above? ☝️

As I was reading what you wrote I was thinking about transparent borders. 

This solution helps accomplish the hover case (which ofc right now is not something we need to worry about in this context, but it's heavy on my mind since I've been struggling with that lately) where we would want the solid button to appear as the outline version while hovering without having to tinker with the padding. I like it very much from a themer's perspective. 

As for the user experience, I guess it's not ideal. It would make sense to me that border width of the button would stay consistent regardless of the variation we are editing, I don't see why the value would be different between them.

**Comment by jasmussen** [comment]
> As for the user experience, I guess it's not ideal. It would make sense to me that border width of the button would stay consistent regardless of the variation we are editing, I don't see why the value would be different between them.

It would certainly start out the same, but I imagine it would still make sense for the border width to be a per-button property. 

Of the options on deck, and since you seem mostly on board with the transparent border idea (thanks for looking), to me it seems like the best path forward so we can replace that padding calc-value. What do you think, @kjellr? I know we discussed this in the past.

**Comment by mcsf** [comment]
> I like this direction personally. Should we also update the site editor style rendered?
> Curious to get more opinions on this one. 

I like this too. 

This PR also highlights the need for a way to add comments to style configurations. The 9999px radius is a great example of this. IIRC this feedback was also recently relayed by Anne from the FSE testing group.

**Comment by oandregal** [comment]
Ok, took a deeper look at this and I believe we need to do a couple of things first.

This is how CSS bundled with the block library behaves:

- It's bundled in all editors as well as the front-end for all themes (with and without `theme.json`). The global stylesheet isn't: it's only available for themes with `theme.json` support or the legacy `experimental-link-color` support.

- Some blocks may have additional `theme.scss` styles that are opt-in for themes. I believe it'll be fine to absorb styles in this stylesheet as well, even when the theme didn't opt-in into them: this API will allow them to overwrite them selectively anyway.

- There's also the opt-in flag `should_load_separate_core_block_assets` that themes can use to only enqueue the CSS of blocks in use. The global stylesheet doesn't respect this setting.

- Some people unregister the block-library CSS handle to bundle their own. They could do the same for global styles.

To respect backward compatibility, it seems we have to do the following before we can land a PR like this:

1. Make the global stylesheet available for all themes https://github.com/WordPress/gutenberg/issues/34328 
	- If the theme declares a `theme.json` it works as today.
	- If the theme doesn't declare a `theme.json` we will still enqueue some CSS: the core presets (needed by patterns) and the block styles defined via this new API. These are styles that we are now enqueuing via the block-library anyway. It's not a change of behavior, it's a change in the handle that contains these styles. We shouldn't enqueue the own theme presets declared in functions.php, though: themes that don't have a theme.json are supposed to do this themselves, and enqueuing ours can conflict with theme's styles.

2. Use `should_load_separate_core_block_assets` in the global stylesheet to only enqueue the CSS of blocks in use. https://github.com/WordPress/gutenberg/issues/31642

**Comment by oandregal** [comment]
By the way, I don't have an answer for the lack of comments in JSON or that we don't have access to the SASS variables in the block.json either. I presume that merging block styles with the theme styles still outweighs these concerns.

**Comment by oandregal** [comment]
One of the PRs to unblock this is ready for review at https://github.com/WordPress/gutenberg/pull/34334

File: lib/class-wp-theme-json-resolver-gutenberg.php
Comment (adamziel) [inline lib/class-wp-theme-json-resolver-gutenberg.php]: Would this make more sense?

```suggestion
		$result->merge( self::get_core_data() );
		$result->merge( self::get_block_data() );
```

Comment (getdave) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: In determining the best ordering here we should consider the current loading order of stylesheets. If we start moving blocks to `blockStyles` and these styles start getting added to the global styles CSS _before_ the Core `theme.json` then we could get anomalies. This is because previously block stylesheets would be loaded after the Global Styles CSS file. This could cause some selectors to be overidden.

**Comment by adamziel** [comment]
I've rebased and added a support for comments as follows:

```json
"border": {
	"//": "100% causes an oval, but any explicit but really high value retains the pill shape.",
	"radius": "9999px"
}
```

Anything with the `//` key is then removed in `get_block_data()`.

This still does not support SCSS variables, but I don't think it needs to in the first version. Let's settle for CSS variables for now and reiterate if needed.

cc @scruffian @getdave @draganescu for reviews. We miss docstrings and maybe a few unit tests here, but can you see any fundamental blockers?

File: lib/class-wp-theme-json-resolver-gutenberg.php
Comment (adamziel) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: I've switched the two, there are no conflicts as of today and blocks_data after core_data seems to have more sense.

File: packages/block-library/src/button/block.json
Comment (scruffian) [inline packages/block-library/src/button/block.json:95]: We should document this approach.

File: packages/block-library/src/button/block.json
Comment (scruffian) [inline packages/block-library/src/button/block.json]: I'm not sure about this name. Block styles already means something else. Could this just be `styles` to match the name in theme.json?

**Comment by scruffian** [comment]
Left a couple of minor comments but I'd say this is almost there.

In terms of colors, I think we should investigate adding a few more defaults to the core theme.json, and then finding ways to use those in blocks. So for example the core theme.json could define the defaults for page color and background, and then the button block could use those (but inverted). That shouldn't stop this PR from being merged, but I think it's an important problem to solve that we don't yet have a solution for. That shouldn't block this from merging though.

**Comment by scruffian** [comment]
> In terms of colors, I think we should investigate adding a few more defaults to the core theme.json, and then finding ways to use those in blocks.

This is discussed in depth here: https://github.com/WordPress/gutenberg/pull/39432

File: packages/block-library/src/button/block.json
Comment (draganescu) [reply packages/block-library/src/button/block.json:95]: This is a "cool" feature that should work across all the JSON files used in WordPress, not only for CSS in JSON.

**Comment by getdave** [comment]
Is this still a prototype? I think it's basically there right?

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (getdave) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:172]: This is looking better. Core is the lowest level, then what the blocks add and finally what the Theme defines.

As long as, for blocks opting into this functionality, the ordering of the output CSS is _the same_ as it was prior to this PR then we should be good.

File: packages/block-library/src/button/block.json
Comment (getdave) [inline packages/block-library/src/button/block.json:89]: I know this probably isn't what anyone wants to hear, but should we consider splitting out the implementation details from the concrete usage on the Button block?

That way we can incrementally introduce the capability to new blocks one by one and check and merge PRs as we go.

File: packages/block-library/src/button/block.json
Comment (getdave) [inline packages/block-library/src/button/block.json:108]: If this comment syntax is now a "thing" should we:

- document it
- check that it also works in `theme.json`

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (getdave) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: Should `removeJsonComments` also be applied to `get_core_data()` and `get_theme_data()`?

**Comment by cbirdsong** [comment]
> In terms of colors, I think we should investigate adding a few more defaults to the core theme.json, and then finding ways to use those in blocks. So for example the core theme.json could define the defaults for page color and background, and then the button block could use those (but inverted).

I really like this idea, and it's broadly similar to a system I use in my non-theme.json agency themes. However, it works best if you define more than just the text/background color. I'm going to elaborate on that over in https://github.com/WordPress/gutenberg/issues/39372.

This sort of system is a key part of https://github.com/WordPress/gutenberg/issues/38998, and the idea was also already discussed in https://github.com/WordPress/gutenberg/issues/29568 and https://github.com/WordPress/gutenberg/issues/33180.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (draganescu) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:172]: This reminds me of my own PR [here](https://github.com/WordPress/gutenberg/pull/40656#discussion_r861631990) where the order of merging settings seems to be weird.

File: packages/block-library/src/button/block.json
Comment (gziolo) [reply packages/block-library/src/button/block.json]: The only issue with `styles` is that it already exists as the top-level property [`styles`](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#block-styles) that keeps the list of block style variations. However, this might be a good opportunity to discuss whether the proposed `supports.blockStyles` could be part of block style variations.

**Comment by gziolo** [comment]
We already have a way to define styles with the global `theme.json`:

https://github.com/WordPress/wordpress-develop/blob/15deb80e504a6ede868bedc1e64e448536d88771/src/wp-includes/theme.json#L242-L244

```json
	"styles": {
		"spacing": { "blockGap": "24px" }
	}
```

and also that can be controlled in a more granular way by a theme, example for Twenty Twenty-Two theme:

https://github.com/WordPress/wordpress-develop/blob/15deb80e504a6ede868bedc1e64e448536d88771/src/wp-content/themes/twentytwentytwo/theme.json#L224-L350

a subset of configuration relevant to the Button block updates in this PR looks like:

```json
	"styles": {
		"blocks": {
			"core/button": {
				"border": {
					"radius": "0"
				},
				"color": {
					"background": "var(--wp--preset--color--primary)",
					"text": "var(--wp--preset--color--background)"
				},
				"typography": {
					"fontSize": "var(--wp--preset--font-size--medium)"
				}
			}
		}
	}
```

---

> I've noticed that many of the block styles are attached to classes (has-background, style variations, etc.) so I presume, initially, we won't see a big reduction of CSS. It'll be a good first step to absorb more later (e.g.: [block variations](https://github.com/WordPress/gutenberg/issues/33232)).

Now, it's important to emphasize that we already have a few existing ways of handling CSS code with [block styles](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-styles.md).

Through `block.json` with [block styles](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#block-styles) that I mentioned in https://github.com/WordPress/gutenberg/pull/34180#discussion_r883356786, an example from this PR:

https://github.com/WordPress/gutenberg/blob/f8568876be3fb36f218e5c0266b73e66aa67fe55/packages/block-library/src/button/block.json#L117-L120

Plugin authors can register also block styles with JS, for example:

```js
wp.blocks.registerBlockStyle( 'core/button', {
	name: 'blue-button',
	label: 'Blue Button',
} );
```

but also with PHP, for example:

```php
register_block_style(
    'core/button',
    array(
        'name'         => 'blue-button',
        'label'        => __( 'Blue Button', 'textdomain' ),
        'inline_style' => '.wp-block-button-style-blue-button { color: blue; }',
    )
);
```

As of today block styles (variations) operate as external or inline stylesheets that contain CSS rules that use the class name of the block (e.g. `wp-block-quote`) or/and the name of the style (e.g. `is-style-blue-quote`).

My question would be whether you can think of a way to integrate the proposed `supports.blockStyles` object into the top-level `styles` array of objects and API methods to register new block styles. I'm aware that the biggest challenge is that currently, you can apply only a single block style at a time. That means in the most naive implementation, we would have to duplicate most of the style settings between all block style definitions. However, maybe there could be a nicer way to do so in the future

In effect, we could have a unified API that integrates seamlessly with what `theme.json` already supports:

```json
"styles": [
 	{
 	 	"name": "blue",
 	 	"label": "Blue Button",
 	 	"settings": {
 	 	 	"color": {
 	 	 	 	"text": "blue"
 	 	 	}
 	 	}
 	}
]
```

or

```js
wp.blocks.registerBlockStyle( 'core/button', {
	name: 'blue-button',
	label: __( 'Blue Button', 'textdomain' ),
        settings: {
              "color": {
			"text": "blue"
		},
       },
} );
```

or

```php
register_block_style(
    'core/button',
    array(
        'name'    => 'blue-button',
        'label'      => __( 'Blue Button', 'textdomain' ),
        'settings' => array(
                'color' => array(
                    'text' => 'blue',
                ),
        ),
    )
);
```

**Comment by gziolo** [comment]
I was thinking even more about existing APIs in the `block.json` definition. There is also [`style`](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#style), example for the Block button:

https://github.com/WordPress/gutenberg/blob/f8568876be3fb36f218e5c0266b73e66aa67fe55/packages/block-library/src/button/block.json#L122

It's a style handle for the registered file or inline styles. It also can be a file reference to an existing file, or even multiple entries, for example:

```json
"style": [ "file:./style.css", "my-style-handle" ]
```

We could always allow including an object there in the array, so using the same use case from my previous comment, we would have for the Block button:

```json
"style": [ "wp-block-button", { "color": { "text": "blue" } } ]
```

**Comment by scruffian** [comment]
> We already have a way to define styles with the global theme.json:

IMO it's better to store this stuff in block.json as it keeps the different parts of the block together in the same files. What do you think @gziolo?

**Comment by scruffian** [comment]
> I was thinking even more about existing APIs in the block.json definition. There is also [style](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#style), example for the Block button

I like this idea.

**Comment by scruffian** [comment]
> My question would be whether you can think of a way to integrate the proposed supports.blockStyles object into the top-level styles array of objects and API methods to register new block styles.

I like this idea too, but I'm not sure we need to do it before we merge this, do we?

**Comment by getdave** [comment]
> We already have a way to define styles with the global theme.json:

Yes but that file is for Theme developers to add block specific styles. What this PR is proposing is codifying the way block's _themselves_ register their styles.

By using a `.json` format we gain all the benefits of programmatic style generation. It should also make it easier to manage specificity by making the inclusion order:

* Core styles
* Block (this PR) styles
* Theme styles
* User styles

I think perhaps what you are saying is how do/are we ensuring that we merge the styles from `block.json` with any block specific styles coming from `theme.json`?

**Comment by scruffian** [comment]
> Yes but that file is for Theme developers to add block specific styles

There is also the core theme.json file...

**Comment by scruffian** [comment]
I added ca5caf70217f21158d9f3763b5699344d449af3a, which moves the styles to the `style` key.

File: lib/compat/wordpress-6.1/blocks.php
Comment (getdave) [inline lib/compat/wordpress-6.1/blocks.php]: What line is this? Can you include an explicit reference? Ideally I won't have to parse the code and learn context to appreciate what this does.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (getdave) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:106]: This should have a docblock because it's `public`.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (getdave) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:172]: > ...seems to be weird.

Can you clarify whether you are raising a concern here that should be addressed? 🙇

**Review by getdave — COMMENTED** [review]
I can't manually test this because I'm getting

> Warning: strpos() expects parameter 1 to be string, array given in /var/www/html/wp-includes/blocks.php on line 20 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-includes/blocks.php:20) in /var/www/html/wp-includes/pluggable.php on line 1421 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-includes/blocks.php:20) in /var/www/html/wp-includes/pluggable.php on line 1424

I think we have a problem with `_gutenberg_multiple_block_styles`.

File: lib/compat/wordpress-6.1/blocks.php
Comment (getdave) [reply lib/compat/wordpress-6.1/blocks.php]: Here's the ref to the line that was removed. 

https://github.com/WordPress/wordpress-develop/blob/6dab53e93bb948a59e48b5856208946d9ee83e74/src/wp-includes/blocks.php#L1267

```php
// Only return the 1st item in the array.
$metadata[ $key ] = $default_style;
```

**Comment by scruffian** [comment]
> I added https://github.com/WordPress/gutenberg/commit/ca5caf70217f21158d9f3763b5699344d449af3a, which moves the styles to the style key.

I reverted this so we could keep this simple. The attempt is moved to here: https://github.com/WordPress/gutenberg/pull/41656

File: packages/block-library/src/button/block.json
Comment (adamziel) [reply packages/block-library/src/button/block.json]: Would be great to do that! @scruffian, @getdave, and I explored that today and it turned out to be quite an effort. Let's make the `blockStyles` key experimental and merge this PR to unblock other work, then let's find a way to bring it to the top-level styles property.

**Review by adamziel — APPROVED** [review]

File: packages/block-library/src/button/block.json
Comment (ajlende) [reply packages/block-library/src/button/block.json:95]: Related to https://github.com/WordPress/gutenberg/issues/35099

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php
Comment (draganescu) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:172]: I don't think I had a concern ..

File: packages/block-library/src/button/block.json
Comment (getdave) [reply packages/block-library/src/button/block.json:95]: @scruffian You should consider splitting this commenting feature out and making it work on all of `theme.json`.

Comment (scruffian) [reply packages/block-library/src/button/block.json:95]: The feedback in #35099 seems to be that this isn't the best way to do it.

**Comment by mburridge** [comment]
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.

**Comment by scruffian** [comment]
We posted about this here: https://make.wordpress.org/core/2022/08/01/moving-core-block-styling-to-json/

**Comment by mburridge** [comment]
Thanks @scruffian, might it be worth adding the `#dev-notes` tag to that post?

**Comment by scruffian** [comment]
Done, thanks

**Comment by scruffian** [comment]
Draft dev note:

## Why
Previously blocks styles have be expressed in CSS. When themes want to override these styles using theme.json, it can be difficult to ensure that the specificity of the CSS that Global Styles generates overrides the default block CSS. This struggle has resulted in the use of some convoluted selectors in block CSS and also some issues where Global Styles settings weren't working as expected.

## What
To overcome this it is now possible to express styles for a block in JSON, using the same properties as a theme.json file. These styles are merged with the Global Styles settings, with theme and user style settings taking priority over block styles. This means that blocks which express their styles in JSON will no longer have to deal with specificity battles between their CSS and Global Styles. In addition, when themes and users override these rules then no extra CSS will be output, which has a performance benefit.

## How
Blocks can add an `__experimentalStyle` property to their block.json file which defines style for the block. For example the pullquote block sets the following:

```
		"__experimentalStyle": {
			"typography": {
				"fontSize": "1.5em",
				"lineHeight": "1.6"
			}
		}
```

This will output the following CSS in the frontend:

```
.wp-block-pullquote{font-size: 1.5em;line-height: 1.6;}
```

**Comment by oandregal** [comment]
There is now https://github.com/WordPress/gutenberg/issues/45198 to track the next steps for this experimental API.

**Pull request merged**
